File: testdata/fixtures/clojure/CWE-327.clj
;; CWE-327: Use of a Broken or Risky Cryptographic Algorithm
;; MD5 is broken and should not be used

(ns fixture.cwe327)

(defn vuln-md5 [input]
  ;; ruleid: use-of-md5
  (MessageDigest/getInstance "MD5")
  ;; ruleid: use-of-md5
  (java.security.MessageDigest/getInstance "MD5")
  ;; ruleid: use-of-md5
  (MessageDigest/getInstance MessageDigestAlgorithms/MD5)
  input)

(defn safe-md5 [input]
  ;; ok: use-of-md5
  (MessageDigest/getInstance "SHA-256")
  ;; ok: use-of-md5
  (java.security.MessageDigest/getInstance "SHA-256")
  input)

File: testdata/fixtures/java/CWE-327.java
package fixture;

// CWE-327: Use of a Broken or Risky Cryptographic Algorithm - MD5/SHA1

import java.security.MessageDigest;

class Cwe327WeakHash {

    byte[] md5(byte[] data) throws Exception {
        // ruleid: java-weak-hash
        MessageDigest md = MessageDigest.getInstance("MD5");
        return md.digest(data);
    }

    byte[] sha1(byte[] data) throws Exception {
        // ruleid: java-weak-hash
        MessageDigest md = MessageDigest.getInstance("SHA1");
        return md.digest(data);
    }

    byte[] sha256(byte[] data) throws Exception {
        // ok: java-weak-hash
        MessageDigest md = MessageDigest.getInstance("SHA-256");
        return md.digest(data);
    }

    byte[] sha512(byte[] data) throws Exception {
        // ok: java-weak-hash
        MessageDigest md = MessageDigest.getInstance("SHA-512");
        return md.digest(data);
    }
}
    byte[] md5Again(byte[] data) throws Exception {
        // ruleid: java-weak-hash
        MessageDigest md = MessageDigest.getInstance("MD5");
        return md.digest(data);
    }

    byte[] sha1Again(byte[] data) throws Exception {
        // ruleid: java-weak-hash
        MessageDigest md = MessageDigest.getInstance("SHA-1");
        return md.digest(data);
    }

    byte[] sha256b(byte[] data) throws Exception {
        // ok: java-weak-hash
        MessageDigest md = MessageDigest.getInstance("SHA-384");
        return md.digest(data);
    }
